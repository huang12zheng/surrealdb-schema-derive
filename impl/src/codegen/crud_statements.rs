use proc_macro2::TokenStream;
use quote::quote;
use syn::Ident;

pub(crate) fn gen_fn_fetch_from_datastore(table_name: &Ident) -> TokenStream {
    quote! {
        async fn fetch_from_datastore(id: u64) -> surrealdb_schema_derive::anyhow::Result<Option<Self::Row>> {
            let query = surrealdb_schema_derive::runtime::make_lookup_query(
                stringify!(#table_name).into(),
                surrealdb_schema_derive::surrealdb::sql::Id::from(id),
            );
            let db = DBX.get().unwrap();

            Ok(surrealdb_schema_derive::runtime::run_single_row_query(&db.datastore, &db.session, query).await?)
        }
    }
}

pub(crate) fn gen_fn_save_to_datastore(table_name: &Ident) -> TokenStream {
    quote! {
        async fn save_to_datastore(self) -> surrealdb_schema_derive::anyhow::Result<Self::Row> {
            let query = surrealdb_schema_derive::surrealdb::sql::Query(surrealdb_schema_derive::surrealdb::sql::Statements(vec![
                surrealdb_schema_derive::surrealdb::sql::Statement::Insert(surrealdb_schema_derive::surrealdb::sql::statements::InsertStatement {
                    into: surrealdb_schema_derive::surrealdb::sql::Table(stringify!(#table_name).into()),
                    data: surrealdb_schema_derive::runtime::to_single_row_values_expression(self.into())?,
                    ignore: false,
                    update: None,
                    output: Some(surrealdb_schema_derive::surrealdb::sql::Output::After),
                    timeout: None,
                    parallel: false,
                }),
            ]));
            let db = DBX.get().unwrap();
            let row = surrealdb_schema_derive::runtime::run_single_row_query(&db.datastore, &db.session, query).await;
            Ok(row?.unwrap())
        }
    }
}

pub(crate) fn gen_fn_delete_to_datastore(table_name: &Ident) -> TokenStream {
    quote! {
        async fn delete_to_datastore(id: u64) -> bool {
            let query = surrealdb_schema_derive::surrealdb::sql::Query(surrealdb_schema_derive::surrealdb::sql::Statements(vec![
                surrealdb_schema_derive::surrealdb::sql::Statement::Delete(surrealdb_schema_derive::surrealdb::sql::statements::DeleteStatement {
                    what: Values(vec![Value::Thing(Thing {
                        tb: stringify!(#table_name).into(),
                        id: surrealdb_schema_derive::surrealdb::sql::Id::from(id),
                    })]),
                    cond: None,
                    output: None,
                    timeout: None,
                    parallel: false,
                }),
            ]));
            let db = DBX.get().unwrap();

            surrealdb_schema_derive::runtime::run_single_row_query::<Self>(&db.datastore, &db.session, query).await.is_ok()
        }
    }
}
