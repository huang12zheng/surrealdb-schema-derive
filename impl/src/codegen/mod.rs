pub(crate) mod object_conversion;
