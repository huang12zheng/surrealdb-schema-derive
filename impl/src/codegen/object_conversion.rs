use proc_macro2::TokenStream;
use quote::quote;
use syn::{FieldsNamed, GenericArgument, Ident, PathArguments, Type};

pub(crate) fn gen_into_value(struct_ident: &Ident, fields: &FieldsNamed) -> TokenStream {
    let field_conversions: Vec<TokenStream> = fields
        .named
        .iter()
        .map(|field| {
            let field_ident = field.ident.clone().unwrap();
            let field_ref = if maybe_extract_optional(field).is_some() {
                quote! {surrealdb_obj_derive::SurrealOption(self.#field_ident)}
            } else {
                quote! {self.#field_ident}
            };
            quote! {
                (stringify!(#field_ident).into(), {
                    let surreal_value: surrealdb_obj_derive::surrealdb::sql::Value = #field_ref.into();
                    surreal_value
                })
            }
        })
        .collect();

    quote! {
        impl Into<surrealdb_obj_derive::surrealdb::sql::Value> for #struct_ident {
            fn into(self) -> surrealdb_obj_derive::surrealdb::sql::Value {
                surrealdb_obj_derive::surrealdb::sql::Value::Object(
                    surrealdb_obj_derive::surrealdb::sql::Object(std::collections::BTreeMap::from([
                        #(#field_conversions),*
                    ]))
                )
            }
        }
    }
}

fn maybe_extract_optional(field: &syn::Field) -> Option<Type> {
    if let Type::Path(ref path_type) = field.ty {
        if let Some(first) = path_type.path.segments.first() {
            if first.ident == "Option" {
                if let PathArguments::AngleBracketed(angle_bracketed) = &first.arguments {
                    let first_arg = angle_bracketed.args.first();
                    if let Some(GenericArgument::Type(inner_type)) = first_arg {
                        Some(inner_type.clone())
                    } else {
                        panic!("Invalid option: {:?}", path_type);
                    }
                } else {
                    panic!("Invalid option: {:?}", path_type);
                }
            } else {
                None
            }
        } else {
            None
        }
    } else {
        None
    }
}
