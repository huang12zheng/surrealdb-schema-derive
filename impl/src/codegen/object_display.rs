use proc_macro2::TokenStream;
use quote::quote;
use syn::Ident;

pub(crate) fn gen_display_object(struct_ident: &Ident) -> TokenStream {
    quote! {
        impl core::fmt::Display for  #struct_ident {
            fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
                write!(f, "{}", Into::<SurrealValue>::into(self.clone()).0)
            }
        }
    }
}
