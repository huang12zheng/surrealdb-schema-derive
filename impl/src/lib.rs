//! This crate implements the macro for `surrealdb_obj_derive` and should not be used directly.

mod codegen;
mod errors;
pub mod runtime;

use codegen::*;
use derive_builder::Builder;
pub use errors::*;
use proc_macro2::TokenStream;
use quote::quote;
pub use runtime::surreal_value_primitives::*;
pub use runtime::surreal_value_primitives::{SurrealOption, SurrealValue};
use surrealdb::{self, Datastore, Session};
use syn::{parse2, spanned::Spanned, Data, DeriveInput, Fields, FieldsNamed, Ident};

#[derive(Clone)]
pub struct DefineTableContext {
    pub namespace: String,
    pub database: String,
}

#[derive(Builder)]
pub struct DefineTableArgs<'a> {
    pub datastore: &'a Datastore,
    pub session: &'a Session,
    #[builder(setter(into, strip_option), default)]
    pub context: Option<DefineTableContext>,
}

pub trait SurrealDbObject:
    TryFrom<SurrealValue, Error = SurrealDbSchemaDeriveQueryError> + Into<SurrealValue>
{
    fn get_table_name() -> String;
}

#[doc(hidden)]
pub fn derive_surreal_db_object(_item: TokenStream) -> Result<TokenStream, syn::Error> {
    let (struct_ident, fields) = extract_derive_struct(_item)?;
    gen_surreal_db_object(&struct_ident, &fields)
}

fn gen_surreal_db_object(
    struct_ident: &Ident,
    fields: &FieldsNamed,
) -> Result<TokenStream, syn::Error> {
    let impl_into_surreal_value = object_conversion::gen_into_value(&struct_ident, &fields);
    let impl_display_object = object_display::gen_display_object(&struct_ident);
    return Ok(TokenStream::from(quote! {
        #impl_into_surreal_value
        #impl_display_object

        impl SurrealDbObject for #struct_ident {

            fn get_table_name() -> String {
                stringify!(#struct_ident).into()
            }
        }


    }));
}

fn extract_derive_struct(struct_stream: TokenStream) -> Result<(Ident, FieldsNamed), syn::Error> {
    let top_level_error_span = struct_stream.span();
    let input: DeriveInput = parse2(struct_stream)?;
    let ident = input.ident;
    let fields = (if let Data::Struct(data_struct) = input.data {
        if let Fields::Named(named_fields) = data_struct.fields {
            Ok(named_fields)
        } else {
            Err(syn::Error::new(
                top_level_error_span,
                "Must use named fields",
            ))
        }
    } else {
        Err(syn::Error::new(
            top_level_error_span,
            "Only structs are supported for surrealdb deriving",
        ))
    })?;
    Ok((ident, fields))
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::*;

    #[test]
    fn derives_simple_struct() {
        assert_debug_snapshot!(derive_surreal_db_object(quote! {
            struct MyStruct {
                name: String,
                count: u8,
                value: isize,
            }
        })
        .unwrap());
        assert_debug_snapshot!(derive_surreal_db_object(quote! {
            struct RustStruct {
                name: Option<String>,
                generics: RustGenerics,
            }
        })
        .unwrap());
    }
}
