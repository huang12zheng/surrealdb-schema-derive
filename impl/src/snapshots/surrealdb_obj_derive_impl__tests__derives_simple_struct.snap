---
source: impl/src/lib.rs
expression: "derive_surreal_db_object(quote! {\n                struct MyStruct { name : String, count : u8, value : isize, }\n            }).unwrap()"
---
TokenStream [
    Ident {
        sym: impl,
    },
    Ident {
        sym: Into,
    },
    Punct {
        char: '<',
        spacing: Alone,
    },
    Ident {
        sym: surrealdb,
    },
    Punct {
        char: ':',
        spacing: Joint,
    },
    Punct {
        char: ':',
        spacing: Alone,
    },
    Ident {
        sym: Value,
    },
    Punct {
        char: '>',
        spacing: Alone,
    },
    Ident {
        sym: for,
    },
    Ident {
        sym: MyStruct,
    },
    Group {
        delimiter: Brace,
        stream: TokenStream [
            Ident {
                sym: fn,
            },
            Ident {
                sym: into,
            },
            Group {
                delimiter: Parenthesis,
                stream: TokenStream [
                    Ident {
                        sym: self,
                    },
                ],
            },
            Punct {
                char: '-',
                spacing: Joint,
            },
            Punct {
                char: '>',
                spacing: Alone,
            },
            Ident {
                sym: surrealdb,
            },
            Punct {
                char: ':',
                spacing: Joint,
            },
            Punct {
                char: ':',
                spacing: Alone,
            },
            Ident {
                sym: Value,
            },
            Group {
                delimiter: Brace,
                stream: TokenStream [
                    Ident {
                        sym: surrealdb_obj_derive,
                    },
                    Punct {
                        char: ':',
                        spacing: Joint,
                    },
                    Punct {
                        char: ':',
                        spacing: Alone,
                    },
                    Ident {
                        sym: surrealdb,
                    },
                    Punct {
                        char: ':',
                        spacing: Joint,
                    },
                    Punct {
                        char: ':',
                        spacing: Alone,
                    },
                    Ident {
                        sym: sql,
                    },
                    Punct {
                        char: ':',
                        spacing: Joint,
                    },
                    Punct {
                        char: ':',
                        spacing: Alone,
                    },
                    Ident {
                        sym: Value,
                    },
                    Punct {
                        char: ':',
                        spacing: Joint,
                    },
                    Punct {
                        char: ':',
                        spacing: Alone,
                    },
                    Ident {
                        sym: Object,
                    },
                    Group {
                        delimiter: Parenthesis,
                        stream: TokenStream [
                            Ident {
                                sym: surrealdb_obj_derive,
                            },
                            Punct {
                                char: ':',
                                spacing: Joint,
                            },
                            Punct {
                                char: ':',
                                spacing: Alone,
                            },
                            Ident {
                                sym: surrealdb,
                            },
                            Punct {
                                char: ':',
                                spacing: Joint,
                            },
                            Punct {
                                char: ':',
                                spacing: Alone,
                            },
                            Ident {
                                sym: sql,
                            },
                            Punct {
                                char: ':',
                                spacing: Joint,
                            },
                            Punct {
                                char: ':',
                                spacing: Alone,
                            },
                            Ident {
                                sym: Object,
                            },
                            Group {
                                delimiter: Parenthesis,
                                stream: TokenStream [
                                    Ident {
                                        sym: std,
                                    },
                                    Punct {
                                        char: ':',
                                        spacing: Joint,
                                    },
                                    Punct {
                                        char: ':',
                                        spacing: Alone,
                                    },
                                    Ident {
                                        sym: collections,
                                    },
                                    Punct {
                                        char: ':',
                                        spacing: Joint,
                                    },
                                    Punct {
                                        char: ':',
                                        spacing: Alone,
                                    },
                                    Ident {
                                        sym: BTreeMap,
                                    },
                                    Punct {
                                        char: ':',
                                        spacing: Joint,
                                    },
                                    Punct {
                                        char: ':',
                                        spacing: Alone,
                                    },
                                    Ident {
                                        sym: from,
                                    },
                                    Group {
                                        delimiter: Parenthesis,
                                        stream: TokenStream [
                                            Group {
                                                delimiter: Bracket,
                                                stream: TokenStream [
                                                    Group {
                                                        delimiter: Parenthesis,
                                                        stream: TokenStream [
                                                            Ident {
                                                                sym: stringify,
                                                            },
                                                            Punct {
                                                                char: '!',
                                                                spacing: Alone,
                                                            },
                                                            Group {
                                                                delimiter: Parenthesis,
                                                                stream: TokenStream [
                                                                    Ident {
                                                                        sym: name,
                                                                    },
                                                                ],
                                                            },
                                                            Punct {
                                                                char: '.',
                                                                spacing: Alone,
                                                            },
                                                            Ident {
                                                                sym: into,
                                                            },
                                                            Group {
                                                                delimiter: Parenthesis,
                                                                stream: TokenStream [],
                                                            },
                                                            Punct {
                                                                char: ',',
                                                                spacing: Alone,
                                                            },
                                                            Group {
                                                                delimiter: Brace,
                                                                stream: TokenStream [
                                                                    Ident {
                                                                        sym: let,
                                                                    },
                                                                    Ident {
                                                                        sym: surreal_value,
                                                                    },
                                                                    Punct {
                                                                        char: ':',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: surrealdb,
                                                                    },
                                                                    Punct {
                                                                        char: ':',
                                                                        spacing: Joint,
                                                                    },
                                                                    Punct {
                                                                        char: ':',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: Value,
                                                                    },
                                                                    Punct {
                                                                        char: '=',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: self,
                                                                    },
                                                                    Punct {
                                                                        char: '.',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: name,
                                                                    },
                                                                    Punct {
                                                                        char: '.',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: into,
                                                                    },
                                                                    Group {
                                                                        delimiter: Parenthesis,
                                                                        stream: TokenStream [],
                                                                    },
                                                                    Punct {
                                                                        char: ';',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: surreal_value,
                                                                    },
                                                                ],
                                                            },
                                                        ],
                                                    },
                                                    Punct {
                                                        char: ',',
                                                        spacing: Alone,
                                                    },
                                                    Group {
                                                        delimiter: Parenthesis,
                                                        stream: TokenStream [
                                                            Ident {
                                                                sym: stringify,
                                                            },
                                                            Punct {
                                                                char: '!',
                                                                spacing: Alone,
                                                            },
                                                            Group {
                                                                delimiter: Parenthesis,
                                                                stream: TokenStream [
                                                                    Ident {
                                                                        sym: count,
                                                                    },
                                                                ],
                                                            },
                                                            Punct {
                                                                char: '.',
                                                                spacing: Alone,
                                                            },
                                                            Ident {
                                                                sym: into,
                                                            },
                                                            Group {
                                                                delimiter: Parenthesis,
                                                                stream: TokenStream [],
                                                            },
                                                            Punct {
                                                                char: ',',
                                                                spacing: Alone,
                                                            },
                                                            Group {
                                                                delimiter: Brace,
                                                                stream: TokenStream [
                                                                    Ident {
                                                                        sym: let,
                                                                    },
                                                                    Ident {
                                                                        sym: surreal_value,
                                                                    },
                                                                    Punct {
                                                                        char: ':',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: surrealdb,
                                                                    },
                                                                    Punct {
                                                                        char: ':',
                                                                        spacing: Joint,
                                                                    },
                                                                    Punct {
                                                                        char: ':',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: Value,
                                                                    },
                                                                    Punct {
                                                                        char: '=',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: self,
                                                                    },
                                                                    Punct {
                                                                        char: '.',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: count,
                                                                    },
                                                                    Punct {
                                                                        char: '.',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: into,
                                                                    },
                                                                    Group {
                                                                        delimiter: Parenthesis,
                                                                        stream: TokenStream [],
                                                                    },
                                                                    Punct {
                                                                        char: ';',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: surreal_value,
                                                                    },
                                                                ],
                                                            },
                                                        ],
                                                    },
                                                    Punct {
                                                        char: ',',
                                                        spacing: Alone,
                                                    },
                                                    Group {
                                                        delimiter: Parenthesis,
                                                        stream: TokenStream [
                                                            Ident {
                                                                sym: stringify,
                                                            },
                                                            Punct {
                                                                char: '!',
                                                                spacing: Alone,
                                                            },
                                                            Group {
                                                                delimiter: Parenthesis,
                                                                stream: TokenStream [
                                                                    Ident {
                                                                        sym: value,
                                                                    },
                                                                ],
                                                            },
                                                            Punct {
                                                                char: '.',
                                                                spacing: Alone,
                                                            },
                                                            Ident {
                                                                sym: into,
                                                            },
                                                            Group {
                                                                delimiter: Parenthesis,
                                                                stream: TokenStream [],
                                                            },
                                                            Punct {
                                                                char: ',',
                                                                spacing: Alone,
                                                            },
                                                            Group {
                                                                delimiter: Brace,
                                                                stream: TokenStream [
                                                                    Ident {
                                                                        sym: let,
                                                                    },
                                                                    Ident {
                                                                        sym: surreal_value,
                                                                    },
                                                                    Punct {
                                                                        char: ':',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: surrealdb,
                                                                    },
                                                                    Punct {
                                                                        char: ':',
                                                                        spacing: Joint,
                                                                    },
                                                                    Punct {
                                                                        char: ':',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: Value,
                                                                    },
                                                                    Punct {
                                                                        char: '=',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: self,
                                                                    },
                                                                    Punct {
                                                                        char: '.',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: value,
                                                                    },
                                                                    Punct {
                                                                        char: '.',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: into,
                                                                    },
                                                                    Group {
                                                                        delimiter: Parenthesis,
                                                                        stream: TokenStream [],
                                                                    },
                                                                    Punct {
                                                                        char: ';',
                                                                        spacing: Alone,
                                                                    },
                                                                    Ident {
                                                                        sym: surreal_value,
                                                                    },
                                                                ],
                                                            },
                                                        ],
                                                    },
                                                ],
                                            },
                                        ],
                                    },
                                ],
                            },
                        ],
                    },
                ],
            },
        ],
    },
    Ident {
        sym: impl,
    },
    Ident {
        sym: core,
    },
    Punct {
        char: ':',
        spacing: Joint,
    },
    Punct {
        char: ':',
        spacing: Alone,
    },
    Ident {
        sym: fmt,
    },
    Punct {
        char: ':',
        spacing: Joint,
    },
    Punct {
        char: ':',
        spacing: Alone,
    },
    Ident {
        sym: Display,
    },
    Ident {
        sym: for,
    },
    Ident {
        sym: MyStruct,
    },
    Group {
        delimiter: Brace,
        stream: TokenStream [
            Ident {
                sym: fn,
            },
            Ident {
                sym: fmt,
            },
            Group {
                delimiter: Parenthesis,
                stream: TokenStream [
                    Punct {
                        char: '&',
                        spacing: Alone,
                    },
                    Ident {
                        sym: self,
                    },
                    Punct {
                        char: ',',
                        spacing: Alone,
                    },
                    Ident {
                        sym: f,
                    },
                    Punct {
                        char: ':',
                        spacing: Alone,
                    },
                    Punct {
                        char: '&',
                        spacing: Alone,
                    },
                    Ident {
                        sym: mut,
                    },
                    Ident {
                        sym: core,
                    },
                    Punct {
                        char: ':',
                        spacing: Joint,
                    },
                    Punct {
                        char: ':',
                        spacing: Alone,
                    },
                    Ident {
                        sym: fmt,
                    },
                    Punct {
                        char: ':',
                        spacing: Joint,
                    },
                    Punct {
                        char: ':',
                        spacing: Alone,
                    },
                    Ident {
                        sym: Formatter,
                    },
                    Punct {
                        char: '<',
                        spacing: Alone,
                    },
                    Punct {
                        char: '\'',
                        spacing: Joint,
                    },
                    Ident {
                        sym: _,
                    },
                    Punct {
                        char: '>',
                        spacing: Alone,
                    },
                ],
            },
            Punct {
                char: '-',
                spacing: Joint,
            },
            Punct {
                char: '>',
                spacing: Alone,
            },
            Ident {
                sym: core,
            },
            Punct {
                char: ':',
                spacing: Joint,
            },
            Punct {
                char: ':',
                spacing: Alone,
            },
            Ident {
                sym: fmt,
            },
            Punct {
                char: ':',
                spacing: Joint,
            },
            Punct {
                char: ':',
                spacing: Alone,
            },
            Ident {
                sym: Result,
            },
            Group {
                delimiter: Brace,
                stream: TokenStream [
                    Ident {
                        sym: write,
                    },
                    Punct {
                        char: '!',
                        spacing: Alone,
                    },
                    Group {
                        delimiter: Parenthesis,
                        stream: TokenStream [
                            Ident {
                                sym: f,
                            },
                            Punct {
                                char: ',',
                                spacing: Alone,
                            },
                            Literal {
                                lit: "{}",
                            },
                            Punct {
                                char: ',',
                                spacing: Alone,
                            },
                            Ident {
                                sym: Into,
                            },
                            Punct {
                                char: ':',
                                spacing: Joint,
                            },
                            Punct {
                                char: ':',
                                spacing: Alone,
                            },
                            Punct {
                                char: '<',
                                spacing: Alone,
                            },
                            Ident {
                                sym: SurrealValue,
                            },
                            Punct {
                                char: '>',
                                spacing: Alone,
                            },
                            Punct {
                                char: ':',
                                spacing: Joint,
                            },
                            Punct {
                                char: ':',
                                spacing: Alone,
                            },
                            Ident {
                                sym: into,
                            },
                            Group {
                                delimiter: Parenthesis,
                                stream: TokenStream [
                                    Ident {
                                        sym: self,
                                    },
                                    Punct {
                                        char: '.',
                                        spacing: Alone,
                                    },
                                    Ident {
                                        sym: clone,
                                    },
                                    Group {
                                        delimiter: Parenthesis,
                                        stream: TokenStream [],
                                    },
                                ],
                            },
                            Punct {
                                char: '.',
                                spacing: Alone,
                            },
                            Literal {
                                lit: 0,
                            },
                        ],
                    },
                ],
            },
        ],
    },
    Ident {
        sym: impl,
    },
    Ident {
        sym: SurrealDbObject,
    },
    Ident {
        sym: for,
    },
    Ident {
        sym: MyStruct,
    },
    Group {
        delimiter: Brace,
        stream: TokenStream [
            Ident {
                sym: fn,
            },
            Ident {
                sym: get_table_name,
            },
            Group {
                delimiter: Parenthesis,
                stream: TokenStream [],
            },
            Punct {
                char: '-',
                spacing: Joint,
            },
            Punct {
                char: '>',
                spacing: Alone,
            },
            Ident {
                sym: String,
            },
            Group {
                delimiter: Brace,
                stream: TokenStream [
                    Ident {
                        sym: stringify,
                    },
                    Punct {
                        char: '!',
                        spacing: Alone,
                    },
                    Group {
                        delimiter: Parenthesis,
                        stream: TokenStream [
                            Ident {
                                sym: MyStruct,
                            },
                        ],
                    },
                    Punct {
                        char: '.',
                        spacing: Alone,
                    },
                    Ident {
                        sym: into,
                    },
                    Group {
                        delimiter: Parenthesis,
                        stream: TokenStream [],
                    },
                ],
            },
        ],
    },
]
